### ISSUE TITLE
SIGSEGV on ~LTMWindow() reproducible always

### ISSUE THREAD

**Issue by potuz** [GoldenCheetah/GoldenCheetah#2615]
I've been experiencing this across many different versions. The current report is on 

```
$ goldencheetah --version
GoldenCheetah DEV-V3.5 1708 (3963)
 $ uname -a 
Linux rye 4.9.20-gentoo #16 SMP Mon Jul 3 20:21:46 -03 2017 x86_64 Intel(R) Core(TM) i5-7200U CPU @ 2.50GHz GenuineIntel GNU/Linux
```

Steps to reproduce

1. Open GC and go to Trends View
2. Open any metric trends chart configured to show some min/max (for example Aerobic power)
3. Click on any maximum or minimum on the chart to open a window with the activities for that week.
4. Close the newly open window. 
5. Close GC. 
6. GC segfaults on closeup, so that next time you open it you need to select a rider instead of opening the previous one. 

Runing the coredump on gdb I can get the following stacktrace on the active thread
```
(gdb) bt
#0  0x000000000065f34e in LTMWindow::~LTMWindow() ()
#1  0x000000000065f429 in LTMWindow::~LTMWindow() ()
#2  0x00007fd7e3995e93 in QObjectPrivate::deleteChildren() () from /home/heluani/Qt/lib/libQt5Core.so.5
#3  0x00007fd7ea9ab614 in QWidget::~QWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#4  0x00007fd7eaaf7639 in QStackedWidget::~QStackedWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#5  0x00007fd7e3995e93 in QObjectPrivate::deleteChildren() () from /home/heluani/Qt/lib/libQt5Core.so.5
#6  0x00007fd7ea9ab614 in QWidget::~QWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#7  0x00007fd7ea9ab8e9 in QWidget::~QWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#8  0x00007fd7e3995e93 in QObjectPrivate::deleteChildren() () from /home/heluani/Qt/lib/libQt5Core.so.5
#9  0x00007fd7ea9ab614 in QWidget::~QWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#10 0x00007fd7eaaf7639 in QStackedWidget::~QStackedWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#11 0x00007fd7e3995e93 in QObjectPrivate::deleteChildren() () from /home/heluani/Qt/lib/libQt5Core.so.5
#12 0x00007fd7ea9ab614 in QWidget::~QWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#13 0x00000000005f7949 in HomeWindow::~HomeWindow() ()
#14 0x00007fd7e3995e93 in QObjectPrivate::deleteChildren() () from /home/heluani/Qt/lib/libQt5Core.so.5
#15 0x00007fd7ea9ab614 in QWidget::~QWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#16 0x0000000000b57e39 in ViewSplitter::~ViewSplitter() ()
#17 0x00007fd7e3995e93 in QObjectPrivate::deleteChildren() () from /home/heluani/Qt/lib/libQt5Core.so.5
#18 0x00007fd7ea9ab614 in QWidget::~QWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#19 0x00007fd7eaaf6869 in QSplitter::~QSplitter() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#20 0x00007fd7e3995e93 in QObjectPrivate::deleteChildren() () from /home/heluani/Qt/lib/libQt5Core.so.5
#21 0x00007fd7ea9ab614 in QWidget::~QWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#22 0x00007fd7eaaf7639 in QStackedWidget::~QStackedWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#23 0x00007fd7e3995e93 in QObjectPrivate::deleteChildren() () from /home/heluani/Qt/lib/libQt5Core.so.5
#24 0x00007fd7ea9ab614 in QWidget::~QWidget() () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#25 0x000000000096fe99 in HomeView::~HomeView() ()
#26 0x00000000009695fa in Tab::~Tab() ()
#27 0x0000000000969649 in Tab::~Tab() ()
#28 0x0000000000902640 in MainWindow::removeTab(Tab*) ()
#29 0x0000000000902b3c in MainWindow::closeEvent(QCloseEvent*) ()
#30 0x00007fd7ea9afc78 in QWidget::event(QEvent*) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#31 0x00007fd7eaaad68b in QMainWindow::event(QEvent*) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#32 0x00007fd7ea96894c in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#33 0x00007fd7ea970366 in QApplication::notify(QObject*, QEvent*) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#34 0x00007fd7e396c9d8 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /home/heluani/Qt/lib/libQt5Core.so.5
#35 0x00007fd7ea9ab07d in QWidgetPrivate::close_helper(QWidgetPrivate::CloseMode) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#36 0x00000000008ff82a in MainWindow::closeAll() ()
#37 0x0000000000b4a300 in MainWindow::qt_static_metacall(QObject*, QMetaObject::Call, int, void**) ()
#38 0x00007fd7e3997751 in QMetaObject::activate(QObject*, int, int, void**) () from /home/heluani/Qt/lib/libQt5Core.so.5
#39 0x00007fd7ea9622e2 in QAction::triggered(bool) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#40 0x00007fd7ea9649f0 in QAction::activate(QAction::ActionEvent) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#41 0x00007fd7ea96534f in QAction::event(QEvent*) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#42 0x00007fd7ea96894c in QApplicationPrivate::notify_helper(QObject*, QEvent*) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#43 0x00007fd7ea970366 in QApplication::notify(QObject*, QEvent*) () from /home/heluani/Qt/lib/libQt5Widgets.so.5
#44 0x00007fd7e396c9d8 in QCoreApplication::notifyInternal2(QObject*, QEvent*) () from /home/heluani/Qt/lib/libQt5Core.so.5
#45 0x00007fd7e44c03d6 in QShortcutMap::dispatchEvent(QKeyEvent*) () from /home/heluani/Qt/lib/libQt5Gui.so.5
#46 0x00007fd7e44c0489 in QShortcutMap::tryShortcut(QKeyEvent*) () from /home/heluani/Qt/lib/libQt5Gui.so.5
#47 0x00007fd7e44748b6 in QWindowSystemInterface::handleShortcutEvent(QWindow*, unsigned long, int, QFlags<Qt::KeyboardModifier>, unsigned int, unsigned int, unsigned int, QString const&, bool, unsigned short) () from /home/heluani/Qt/lib/libQt5Gui.so.5
#48 0x00007fd7e448fba4 in QGuiApplicationPrivate::processKeyEvent(QWindowSystemInterfacePrivate::KeyEvent*) () from /home/heluani/Qt/lib/libQt5Gui.so.5
#49 0x00007fd7e4495055 in QGuiApplicationPrivate::processWindowSystemEvent(QWindowSystemInterfacePrivate::WindowSystemEvent*) () from /home/heluani/Qt/lib/libQt5Gui.so.5
#50 0x00007fd7e447077b in QWindowSystemInterface::sendWindowSystemEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/heluani/Qt/lib/libQt5Gui.so.5
#51 0x00007fd7d6e46f60 in ?? () from /home/heluani/Qt/plugins/platforms/../../lib/libQt5XcbQpa.so.5
#52 0x00007fd7dfb3565d in g_main_context_dispatch () from /usr/lib64/libglib-2.0.so.0
#53 0x00007fd7dfb35940 in ?? () from /usr/lib64/libglib-2.0.so.0
#54 0x00007fd7dfb359ec in g_main_context_iteration () from /usr/lib64/libglib-2.0.so.0
#55 0x00007fd7e39c02cf in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/heluani/Qt/lib/libQt5Core.so.5
#56 0x00007fd7e396a9ca in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) () from /home/heluani/Qt/lib/libQt5Core.so.5
#57 0x00007fd7e3973024 in QCoreApplication::exec() () from /home/heluani/Qt/lib/libQt5Core.so.5
#58 0x00000000004b8acb in main ()
(gdb) info threads
  Id   Target Id         Frame 
* 1    Thread 0x7fd7d706bd00 (LWP 18553) 0x000000000065f34e in LTMWindow::~LTMWindow() ()
  2    Thread 0x7fd7a0ff9700 (LWP 18581) pthread_cond_timedwait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  3    Thread 0x7fd7c51ab700 (LWP 18573) 0x00007fd7e26a1513 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
  4    Thread 0x7fd7ceb2d700 (LWP 18558) 0x00007fd7e2697b5d in poll () at ../sysdeps/unix/syscall-template.S:84
  5    Thread 0x7fd7c49aa700 (LWP 18574) 0x00007fd7e2699783 in select () at ../sysdeps/unix/syscall-template.S:84
  6    Thread 0x7fd7c59ac700 (LWP 18571) 0x00007fd7e2697b5d in poll () at ../sysdeps/unix/syscall-template.S:84
  7    Thread 0x7fd7a3fff700 (LWP 18575) pthread_cond_timedwait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  8    Thread 0x7fd7a17fa700 (LWP 18580) pthread_cond_wait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  9    Thread 0x7fd7a2ffd700 (LWP 18577) pthread_cond_wait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  10   Thread 0x7fd789ffb700 (LWP 18586) pthread_cond_wait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  11   Thread 0x7fd788ff9700 (LWP 18588) 0x00007fd7e26a1513 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
  12   Thread 0x7fd7a27fc700 (LWP 18578) 0x00007fd7e26a1513 in epoll_wait () at ../sysdeps/unix/syscall-template.S:84
  13   Thread 0x7fd7897fa700 (LWP 18587) pthread_cond_wait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  14   Thread 0x7fd78a7fc700 (LWP 18585) pthread_cond_timedwait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  15   Thread 0x7fd7d50e1700 (LWP 18557) 0x00007fd7e2697b5d in poll () at ../sysdeps/unix/syscall-template.S:84
  16   Thread 0x7fd7c6ffd700 (LWP 18570) 0x00007fd7e2697b5d in poll () at ../sysdeps/unix/syscall-template.S:84
  17   Thread 0x7fd7a37fe700 (LWP 18576) pthread_cond_timedwait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
  18   Thread 0x7fd78affd700 (LWP 18584) pthread_cond_wait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  19   Thread 0x7fd78b7fe700 (LWP 18583) pthread_cond_wait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  20   Thread 0x7fd7a1ffb700 (LWP 18579) pthread_cond_wait () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
  21   Thread 0x7fd78bfff700 (LWP 18582) __lll_lock_wait_private () at ../sysdeps/unix/sysv/linux/x86_64/lowlevellock.S:95
(gdb) 
```

The generated core is 1.1Gb so I won' t be able to post it but I'll be happy to assist in anyway you request.

**Comment by liversedge** [comment]
Does this only happen when you have previously clicked on a data  point in the chart to open up the overlay pane?

**Comment by potuz** [comment]
Yes as far as I know, this is the only consistent way I have to trigger this crash.

**Comment by potuz** [comment]
Perhaps I should add that I use the window manager to close the overlay pane, as the little "x" in the top right corner does not close it (I can include a screenshot if this last sentence isn't clear).

**Comment by potuz** [comment]
Indeed removing the call to delete popup in ~LTMWindow() solves the problem for me. I haven't run this through valgrind to check if this is a leak, but at first sight it seems to me that QT's garbage collection takes care of freeing popup on the call to deleteChildren. I can't remove the whole destructor (and leave the default anyway) because there's a template somewhere else that casts to ~T and I didn't spend time looking into that. So my diff here is simply
```
diff --git a/src/Charts/LTMWindow.cpp b/src/Charts/LTMWindow.cpp
index 47bd9257..60f26da4 100644
--- a/src/Charts/LTMWindow.cpp
+++ b/src/Charts/LTMWindow.cpp
@@ -308,7 +308,6 @@ LTMWindow::LTMWindow(Context *context) :
 
 LTMWindow::~LTMWindow()
 {
-    delete popup;
 }
 
 void
```

**Comment by potuz** [comment]
Although I can't trigger the crash with the above diff, I don't think that was the original problem, for some reason every destructor gets called twice in my build.

**Comment by potuz** [comment]
OK I spent the day today trying to figure out this one and I found the problem. The user case might be very tiny to hit this problem: Linux with a windows manager like i3 and a high dpi monitor so that it's unconfortable to hit the "x" button on the top right corner. The problem is that GcPane does not handle the WM_DELETE_WINDOW protocol, for example by defining GcPane::closeEvent. Therefore if a user like myself closes the popup windows with the window manager by sending the above mentioned signal, the popup hides and destroys itself immediately. Then later on exit the GC calls the destructor on the LTMWindow and this deletes explicitly the popup. 

Since the popup is created without any parents, removing the deletes as I  was doing above removes the crash, but its a leak because Qt will not call popup's destructor. 

I think the easiest way to solve this issue is simply by removing the property WA_DeleteOnClose from GcPane as by
```
diff --git a/src/Charts/GcPane.cpp b/src/Charts/GcPane.cpp
index 6948d34e..4115331b 100644
--- a/src/Charts/GcPane.cpp
+++ b/src/Charts/GcPane.cpp
@@ -22,7 +22,6 @@ GcPane::GcPane() : QWidget(NULL, Qt::FramelessWindowHint),
     borderWidth(4), dragState(None)
 {
     closeImage = QPixmap(":images/toolbar/popbutton.png");
-    setAttribute(Qt::WA_DeleteOnClose);
     setAttribute(Qt::WA_TranslucentBackground);
     setAttribute(Qt::WA_NoSystemBackground);
     setWindowFlags(windowFlags() | Qt::WindowStaysOnTopHint);
```
The problem with this approach is that since GcPanes are defined with NULL as parent widgets I do not know if other instances are explicitly deleted when out of scope or closed. So while this will fix in the situation described in this issue, it may create other leaks elsewhere. 

After a quick browse of the code I  didn't see any problems with this so I have now my own copy of GC compiled with this patch.
